### PR TITLE
docs(codes): Add (key) codes documentation

### LINKED ISSUE #218: docs: Codes (Keys & Controls)
Compliments #21.

### PR DESCRIPTION
**Pull request by innovaker**
- Adds documentation (#218) for standardized HID codes (#21).
- Compliments PR #236.
- Can be previewed here: https://deploy-preview-308--zmk.netlify.app/docs/codes/

### Features / Notes
- Mostly data-driven (React components) to ease maintenance, minimize boilerplating and maximize future potential.
- Data components embedded into mdx pages to ease reorganisation/refactoring and add contextual notes.
- Includes a bespoke table footnotes system that leverages mdx.
- Is statically prerendered as part of the docusaurus build, but runs as React components in the browser.
- This is the 3rd iteration, although it's the first to be publicly released 👍 
  - The 1st was mostly HTML = maintenance nightmare! 👎 
  - The 2nd was a hybrid with some components = uncomfortable / messy 👎 
  - I also experimented with markdown tables but didn't like it, and it was a maintenance nightmare. 👎 👎 

### Issues
- Closes #218.

### Testing
- `eslint . --ext js,jsx,md,mdx`
- `npx prettier --write .`
- `npm run build` and `npm run serve`
- visually checked pages for mistakes and oversights
- checked for console errors
- tested multiple footnotes

### Future
- Upgrade to TypeScript.
- Add realtime filtering features (specification in progress).
- Add expandable detail for each item (specification in progress).
- Test each code on each OS and update documentation data (_community effort_!!!).
- Explore generic/adaptive names for common uses such as `VOL_UP` (after OS testing).
- Explore expanding Codes documentation beyond HID (other behaviors).
- Explore potential benefits of moving data outside of docs (perhaps a different format) for use throughout ZMK systems as a single source of truth.

### Feedback / Reviews
- I welcome ideas or feedback!
- This was a big piece of work with lots of data involved, so I expect there to be bugs/oversights!
  - Please report what you spot so that it can be corrected before it affects everyone else!
- We're still open to discussions around key code names, but there has already been a 5+ week RFC process so please check that first (see #21).
- I've never flexed my React skills outside of my personal projects, so I welcome constructive criticism and technical reviews!

### CODE REVIEW

File: docs/src/css/codes.css
Comment (Nicell) [inline docs/src/css/codes.css:38]: This cursor should only be applied to the `a` tag. A simple `> a` should suffice.

File: docs/src/css/codes.css
Comment (Nicell) [inline docs/src/css/codes.css:8]: I'm not a fan of this cursor. I think it should be replaced in favor of the simple `pointer`. On top of this, clicking isn't obvious as to what it does. We may want a toast triggered on click. This needs to be investigated.

Possibly the addition of `user-select: none` would be nice to also give the impression that it's a copy button (along with a toast).

File: docs/src/css/codes.css
Comment (Nicell) [inline docs/src/css/codes.css:12]: Not a fan of making the background seemingly disappear when hovering. Either it should go slightly darker, or just not change at all IMO

File: docs/src/css/codes.css
Comment (Nicell) [inline docs/src/css/codes.css:187]: Is this footnote stuff used? I can't seem to find it...

File: docs/src/css/codes.css
Comment (Nicell) [inline docs/src/css/codes.css:123]: Small aesthetic thing. I'd prefer if these colors were a bit more pastel and subtle. They _feel_ out of place.

File: docs/src/components/codes/OsSupportIcon.jsx
Comment (Nicell) [inline docs/src/components/codes/OsSupportIcon.jsx:35]: These icons work for ok, but I kind of like the aesthetic of the emojis used on our feature table. Thoughts?

File: docs/src/components/codes/FootnoteRefs.jsx
Comment (Nicell) [inline docs/src/components/codes/FootnoteRefs.jsx:24]: I'm curious why this is needed?

**Review by Nicell — COMMENTED** [review]
This overall looks great. The components are split out quite a bit, which should help for making minor adjustments. Most comments are on styling and aesthetics, all in the eye of the beholder. 

Besides that I have these general comments:

- The OS labels are not great IMO. I'd really like to have some icons. W, L, A, m, and i mean nothing until you dig a bit.
- While the compact spacing of the tables lets you see more, it feels claustrophobic and against the grain with the rest of the site's feel. I'd like to see this open up a bit more.
- All of the footnotes code seems to be unused. What did you foresee the use of it as? Does it need to be added to this PR? It is a large portion of this PR.

Thanks again for the work on this! Hope to continue to review this more and get some discussion.

File: docs/src/css/codes.css
Comment (innovaker) [reply docs/src/css/codes.css:38]: Thanks!

File: docs/src/css/codes.css
Comment (innovaker) [reply docs/src/css/codes.css:8]: Heh, agreed.  I'm glad someone noticed. I wasn't a fan either but didn't want to waste time on alternatives.  Will look into it!

File: docs/src/css/codes.css
Comment (innovaker) [reply docs/src/css/codes.css:12]: Yeah, me too.  It was a quick-win solution to cope with both light and dark themes.  I'll look into it further when I get a chance.

File: docs/src/css/codes.css
Comment (innovaker) [reply docs/src/css/codes.css:123]: Agreed, I didn't spend much time on that.  I welcome color suggestions!

File: docs/src/components/codes/OsSupportIcon.jsx
Comment (innovaker) [reply docs/src/components/codes/OsSupportIcon.jsx:35]: Hehe, I actually thought the same ... but about the tick in the features table.  I discussed this on Discord a while back because I'd like a solution that's compatible across different cultures.  The affirmative meaning of the check mark is a very English concept.  It even has the opposite meaning in some cultures.  See https://en.wikipedia.org/wiki/Check_mark for more detail.

I felt the stars were clear enough to be understood without a legend and not too over-stated or under-stated (on the various OS renderings) to unbalance the table.  The 5-star ratings also gave me a giggle.  I also felt that the small cross was more useful than a empty space which could be misinterpreted.  The question mark is pretty self explanatory?

Taking the above into account, any thoughts?

I made sure they all have titles/mouseover hints.

File: docs/src/components/codes/FootnoteRefs.jsx
Comment (innovaker) [reply docs/src/components/codes/FootnoteRefs.jsx:24]: This is to handle multiple footnotes (an array).  It produces a comma delimited list of footnote references.  This seemed like the neatest way to approach it using react elements/fragments.  I welcome a better approach.

**Comment by innovaker** [comment]
Many thanks @Nicell, I appreciate the feedback and suggestions 👍 

---

> The OS labels are not great IMO

I absolutely agree.  This was explored at length on Discord a few weeks ago.  We tried various options.  The current headings were the best of a bad bunch because the OS logos have strict restrictions.  What would be your suggestion please?

---

> While the compact spacing of the tables lets you see more, it feels claustrophobic and against the grain

I agree - to an extent.  I felt that the default spacing and larger fonts made it difficult to visualise/scan the "general vicinity".  They were also problematic for narrower screens.  I welcome suggestions though.  Could you perhaps play around with it to find what you feel are comfortable settings for each resolution band please?

---

> All of the footnotes code seems to be unused. What did you foresee the use of it as? Does it need to be added to this PR? It is a large portion of this PR.

I'm anticipating that the footnotes will be useful as the testing progresses.  In most documentation you'll find footnotes that indicate special cases and they're often per row and per OS.  Because the tables are data-driven, in-built footnote support seemed like a logical approach.

The footnotes are fully functional and tested.  There's also an example footnote that you can experiment with.  To try it:
1. Pull locally
2. `/docs`: `npm ci`
3. Open `docs/src/data/hid.js`
4. Modify your key of choice like this:
```
footnotes: { windows: "example" }
```
5. `/docs`: `npm run start`

---

Thanks!

**Comment by adamcstephens** [comment]
Content wise I like the Codes page. It's clear and easy to navigate. I like the copy on click, but adding the `&kp` into the copied text can add some extra editing when pasting.

Should this PR also update the other pages that reference keys to use the new naming (e.g. the encoder page)?

File: docs/src/components/codes/OsSupportIcon.jsx
Comment (Nicell) [reply docs/src/components/codes/OsSupportIcon.jsx:35]: I think I've found some values I like:

- Windows: #CAEDFD
- Linux: #FFF2CA
- Android: #D8EED9

And then I'm thinking iOS could just be white and MacOS could take iOS's current color.

Result:
image

We also may be able to add `mix-blend-mode: multiply` to `.os`. It keeps the every other row difference in brightness, but it looks terrible in dark mode. If we could turn it off in dark mode, it would be an improvement:
image

File: docs/src/css/codes.css
Comment (Nicell) [reply docs/src/css/codes.css:8]: This looks like a possible good candidate for toasts: https://github.com/caroso1222/notyf

File: docs/src/css/codes.css
Comment (Nicell) [reply docs/src/css/codes.css:123]: I think I've found some values I like:

- Windows: #CAEDFD
- Linux: #FFF2CA
- Android: #D8EED9

And then I'm thinking iOS could just be white and MacOS could take iOS's current color.

Result:
image

We also may be able to add `mix-blend-mode: multiply` to `.os`. It keeps the every other row difference in brightness, but it looks terrible in dark mode. If we could turn it off in dark mode, it would be an improvement:
image

File: docs/src/components/codes/OsSupportIcon.jsx
Comment (Nicell) [reply docs/src/components/codes/OsSupportIcon.jsx:35]: Good points on all of this. Any reason we couldn't use the star, x, and ? emojis? I think they look a bit more lively and in-line with the site's style. Not super important to change that, just a thought!

Comment (innovaker) [reply docs/src/components/codes/OsSupportIcon.jsx:35]: Could.  Which ones did you have in mind?  Emojis seemed quite distracting when I tried some previously.

File: docs/src/css/codes.css
Comment (innovaker) [reply docs/src/css/codes.css:123]: Yeah, I like that.  Presumably it works well enough in dark mode also?

File: docs/src/components/codes/OsSupportIcon.jsx
Comment (Nicell) [reply docs/src/components/codes/OsSupportIcon.jsx:35]: - Star: ⭐
- Cross: ❌
- Question mark: ❔

Maybe they'd be a bit too much, I can see that.

File: docs/src/css/codes.css
Comment (innovaker) [reply docs/src/css/codes.css:123]: Yeah, I like that!  Thanks for looking into it!  I'll give it a go :)

File: docs/sidebars.js
Comment (Nicell) [inline docs/sidebars.js:28]: Just thought of this: Codes might not be super obvious at first that it contains all the defines for HID codes for keys. Maybe Keycodes would be more direct?

Comment (innovaker) [reply docs/sidebars.js:28]: I'm glad you spotted that!  I was hoping someone would bring it up because it was deliberate.  I started off with `Key Codes`, but as time went on and I thought through the bigger picture I realised that:
1. It might be a good section to include summary tables of the non-HID codes also (they can just be mdx pages).  Behavior codes are not currently the easiest to find buried inside the behavior pages, so it might be wise to turn each set of codes into its own mdx page which we can embedded under both sections (mdx pages can be used as components).  This was previously [discussed](https://discordapp.com/channels/719497620560543766/719565084208398406/765590462274469890).
2. We'll soon be moving beyond just key codes.  More generally it's `hid` codes and other behavior codes (which is partly why named the data file `hid.js` instead of `keycodes.js`).  The counterarguments to this weaken somewhat when you start considering encoders, trackballs, mice, sliders, touch controls, etc.
3. We're planning to refactor `keys.h` into `codes.h` in the [near future](https://discordapp.com/channels/719497620560543766/719565084208398406/769222960892346438) so that there's only one point of entry within user configs.
4. `codes` is sometimes used as a shorthand for `key codes` in discussions.

**Comment by joelspadin** [comment]
Unless we can find a better way to list the OSes, I think this needs a legend at the top that says what each letter means. A legend would also be helpful to explain the ⭐✖❔ symbols.

Another idea would be to rotate the text. CSS that results in a reasonably legible headers for me:

```css
.vertical {
  writing-mode: vertical-lr;
  display: block;
  border: none;
  font-weight: 500;
}
```

(I think you'd need to wrap each OS `<th>` in an additional `<span>` though because applying that to multiple header cells just stacks them all vertically in one column.)

Also, in light theme the documentation link icon just looks like a sort of diagonal oval with a circle in the center of it? I couldn't figure out that it was supposed to be chain links until I switched to dark theme.

image

Consider using an icon font where the color can change with the theme and/or using something that's more easily recognizable like the "external link" icon (https://fontawesome.com/icons/external-link-alt).

**Comment by innovaker** [comment]
Thanks for the feedback @joelspadin!

---

> Unless we can find a better way to list the OSes, I think this needs a legend at the top that says what each letter means. A legend would also be helpful to explain the ⭐✖❔ symbols.

That's a good idea, I'll give it a try.

---

> Another idea would be to rotate the text. CSS that results in a reasonably legible headers for me:

I'm personally not a fan because it floods a lot of vertical with whitespace.  I'd prefer a static legend at the top as you suggested above.  Any thoughts @Nicell? 

---

> Consider using an icon font where the color can change with the theme and/or using something that's more easily recognizable like the "external link" icon (https://fontawesome.com/icons/external-link-alt).

Thanks, will look into it!

**Comment by innovaker** [comment]
Today I followed-up on most of the comments.

**The commits must be squashed into the original commit before this PR is merged.**

I welcome further feedback! 👍

File: docs/src/css/codes.css
Comment (innovaker) [reply docs/src/css/codes.css:78]: Please could you the update?

File: docs/src/css/codes.css
Comment (innovaker) [reply docs/src/css/codes.css]: Gave it a shot with a different library.  What do you think?

File: docs/src/css/codes.css
Comment (innovaker) [reply docs/src/css/codes.css]: Went for a more subtle approach.  Any thoughts?

File: docs/src/css/codes.css
Comment (innovaker) [reply docs/src/css/codes.css]: Done.  Any further thoughts?

File: docs/src/components/codes/OsSupportIcon.jsx
Comment (innovaker) [reply docs/src/components/codes/OsSupportIcon.jsx]: Gave it a go!  What do you think?

**Comment by innovaker** [comment]
> Content wise I like the Codes page. It's clear and easy to navigate. I like the copy on click, but adding the `&kp` into the copied text can add some extra editing when pasting.

@adamcstephens, I've updated the clipboard copying as per your feedback.

**Comment by innovaker** [comment]
@joelspadin, I think what I've implemented covers most of your feedback?  Any thoughts please?

File: docs/src/css/codes.css
Comment (Nicell) [reply docs/src/css/codes.css]: Looking much better. I'm thinking that `user-select: none` should be added still. Also on Windows the updated cursor just looks really ugly. Not sure why so many Windows cursors have this old jagged style. Other than that, well done!

Comment (innovaker) [reply docs/src/css/codes.css]: Buy a Mac.

**Comment by joelspadin** [comment]
> @joelspadin, I think what I've implemented covers most of your feedback? Any thoughts please?

Looks good to me!

**Comment by innovaker** [comment]
Please don't merge this issue until I sign it off.

**Comment by petejohanson** [comment]
A few general comments on the end user experience, I will work on also doing a code review here soon:

*Looks great!* I also love the easy ability to copy items right from the docs.

I do notice some odd line breaks in a few areas. Is this intentional?

Screenshot from 2020-11-03 13-16-09

The other note, which doesn't need to be something fixed in this PR, but it might be nice to add some high level summary/description to each major section, to help give some context to the table found in that section.

The other item, it seems like the first column can be completely removed once we merge #300? Again, something we can do quickly next, but will be unneeded soon.

Thanks again.

**Comment by innovaker** [comment]
> Looks great! I also love the easy ability to copy items right from the docs.

Thanks!

---

> I do notice some odd line breaks in a few areas. Is this intentional?

Do you mean the subscripts on the symbols?  If so, that's intentional.  If not, then I'm missing something.

---

> The other note, which doesn't need to be something fixed in this PR, but it might be nice to add some high level summary/description to each major section, to help give some context to the table found in that section.

Agreed.  To be honest, I just thought I'd leave it for someone else.  But that's also why I built the pages in `mdx` rather than `jsx` with a content/wrapper approach (which took *hours* to figure out because the toc system is largely undocumented), so that anybody can add "decoration" text.

---

> The other item, it seems like the first column can be completely removed once we merge #300? Again, something we can do quickly next, but will be unneeded soon.

Precisely.  That was my intent.

**Comment by petejohanson** [comment]
Ok, if the subscript approach was intentional, I can see how that could be helpful, just wasn't expecting it. Thanks!

**Comment by innovaker** [comment]
> Ok, if the subscript approach was intentional, I can see how that could be helpful, just wasn't expecting it. Thanks!

It was pretty impossible to scan read without it.

File: docs/src/data/hid.js
Comment (petejohanson) [inline docs/src/data/hid.js:15]: What is the `item` value here?

**Review by petejohanson — COMMENTED** [review]
Just one question, on a whole, this is great! Thanks for the attention to detail on this.

File: docs/src/data/hid.js
Comment (innovaker) [reply docs/src/data/hid.js:15]: That's the key/control's extended usage value `(page << 16) | id`.  The `application` of said key, is also an extended usage (`KEYBOARD` or `CONSUMER` at the moment, soon to be `MOUSE` as well).

It's how it's described in the HID specification, similar to the terminology we were using in #213.  I've yet to find a better alternative than `item` but I completely understand that it's not clear unless you know HID well.

Comment (innovaker) [reply docs/src/data/hid.js:15]: Also, `usages` is an array to facilitate simultaneous events.  Modified keys is the basic example of this.  But I foresee it being just as applicable to modified mice clicks, or shortcuts, etc.

Comment (petejohanson) [reply docs/src/data/hid.js:15]: No worries. That makes sense, thanks for explaining it.

Comment (innovaker) [reply docs/src/data/hid.js:15]: When I get more time, I might replace those values with their `(page << 16 | id)` equivalents to increase readability.  It was originally JSON which explains why it's in that form currently.

**Comment by innovaker** [comment]
Will tidy this up tomorrow ready for merging.

**Comment by innovaker** [comment]
### Changes
  - Rebased onto `main`.
  - Previous improvements squashed.
  - Further improvements.
    - See commits for details.

### Notes
 - Committed each set of changes individually for the benefit of reviewers.
 - After review, non-conventional commits will be squashed before merging.

**Comment by innovaker** [comment]
I've done some housekeeping and general improvements.  I'd appreciate a quick check/review of the new commits please.

**Review by petejohanson — APPROVED** [review]
LGTM. Would be interesting to get some input from @Nicell on the JS.

**Comment by petejohanson** [comment]
Conflicts from the merge of #300, sorry.

File: docs/src/data/hid.js
Comment (petejohanson) [inline docs/src/data/hid.js:23]: The bit shifting and binary `OR` here seems duplicated, thoughts on making this something like:

```suggestion
        item: hidExtendedUsage(keyPage, 0x04),
```

to parallel that use in the firmware code?

**Review by petejohanson — COMMENTED** [review]
The format for `hid.js` seems very reasonable. I'm glad you're doing `os: null` for untested, having them in there explicitly will make changes smaller and clearer once we test and add things, and easy to search for untested items.

Just one comment on the item encoding.

File: docs/src/data/hid.js
Comment (innovaker) [reply docs/src/data/hid.js]: Please check the re-push.

Comment (petejohanson) [reply docs/src/data/hid.js]: Yeah, looks great!

**Pull request merged**
